firstvoices/backend/models/widget.py

```diff
@@ -0,0 +1,149 @@
+    WIDGET_APPS = 10
+    WIDGET_CONTACT = 20
+    WIDGET_GALLERY = 30
+    WIDGET_IFRAME = 40
+    WIDGET_KEYBOARDS = 50
+    WIDGET_LOGO = 60
+    WIDGET_LIST = 70
+    WIDGET_QUOTES = 80
+    WIDGET_STATS = 90
+    WIDGET_TEXT = 100
+    WIDGET_TEXTCONCISE = 110
+    WIDGET_TEXTFULL = 120
+    WIDGET_TEXTICONS = 130
+    WIDGET_TEXTMULTI = 140
+    WIDGET_WOTD = 150
+
+
+class WidgetFormats(models.IntegerChoices):
+    # enum intentionally has gaps to allow future changes to keep sequential order
+    default = 0
+    left = 10
+    right = 20
+
+
+class Widget(BaseModel):
+    class Meta:
+        verbose_name = _("widget")
+        verbose_name_plural = _("widgets")
+        rules_permissions = {
+            "view": rules.always_allow,
+            "add": predicates.is_superadmin,
+            "change": predicates.is_superadmin,
+            "delete": predicates.is_superadmin,
+        }
+
+    title = models.CharField(max_length=225)
+    type = models.IntegerField(
+        choices=WidgetTypes.choices, default=WidgetTypes.WIDGET_TEXT
+    )
+    format = models.IntegerField(
+        choices=WidgetFormats.choices, default=WidgetFormats.default
+    )
+
+    def __str__(self):
+        return self.title
+
+
+class WidgetSettings(BaseModel):
+    class Meta:
+        verbose_name = _("widget setting")
+        verbose_name_plural = _("widget settings")
+        rules_permissions = {
+            "view": rules.always_allow,
+            "add": predicates.is_superadmin,
+            "change": predicates.is_superadmin,
+            "delete": predicates.is_superadmin,
+        }
+
+    widget = models.ForeignKey(
+        Widget, on_delete=models.CASCADE, related_name="%(class)s_set"
+    )
+    key = models.CharField(max_length=225)
+    value = models.TextField()
+
+    def __str__(self):
+        return self.key
+
+
+class SiteWidget(Widget, BaseControlledSiteContentModel):
+    class Meta:
+        verbose_name = _("site widget")
+        verbose_name_plural = _("site widgets")
+        rules_permissions = {
+            "view": predicates.is_visible_object,
+            "add": predicates.is_superadmin,  # permissions will change when we add a write API
+            "change": predicates.is_superadmin,
+            "delete": predicates.is_superadmin,
+        }
+
+    def __str__(self):
+        return self.title
+
+
+class SiteWidgetList(BaseSiteContentModel):
+    class Meta:
+        verbose_name = _("site widget list")
+        verbose_name_plural = _("site widget lists")
+        rules_permissions = {
+            "view": predicates.has_visible_site,
+            "add": predicates.is_superadmin,  # permissions will change when we add a write API
+            "change": predicates.is_superadmin,
+            "delete": predicates.is_superadmin,
+        }
+
+    widgets = models.ManyToManyField(
+        SiteWidget, related_name="sitewidget_set", through="SiteWidgetListOrder"
+    )
+    title = models.CharField(max_length=225)
```

Comment:
Does a widget list need a title?

Reply:
Good catch, although it can be very hard to tell the lists apart without it, so it may be worth keeping this as an optional field?

Reply:
Would the list's "site" fk not be enough to tell the lists apart? Or will sites have multiple widget lists?

Reply:
A site can have multiple widget lists. The intention is for each page (when the page model is added) on a site to have a SiteWidgetList foreign key relation. A SiteWidgetList also specifies the homepage widgets for a site by a foreign key relation in the site model.

Reply:
Ah I see, thanks for the clarification. Maybe we can nix the title field and when the page model is added we could use the page foreign key to differentiate the lists then?

Reply:
That sounds good, thanks for the feedback.